producer/servicelog/service_log_producer.go

```diff
@@ -0,0 +1,122 @@
+
+// Package servicelog contains functions that can be used to produce (that is
+// send) messages to Service Log.
+package servicelog
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"github.com/RedHatInsights/ccx-notification-service/conf"
+	"github.com/RedHatInsights/ccx-notification-service/types"
+	httputils "github.com/RedHatInsights/insights-operator-utils/http"
+	"github.com/rs/zerolog/log"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+// Producer is an implementation of Producer interface for Service Log
+type Producer struct {
+	Configuration conf.ServiceLogConfiguration
+	AccessToken   string
+}
+
+// New constructs a new instance of Producer implementation
+func New(config conf.ConfigStruct) (*Producer, error) {
+	prod := &Producer{
+		Configuration: conf.GetServiceLogConfiguration(config),
+	}
+	err := prod.refreshToken()
+	if err != nil {
+		return prod, err
+	}
+	return prod, nil
+}
+
+// ProduceMessage sends the given message to Service Log
+func (producer *Producer) ProduceMessage(msg types.ProducerMessage) (partitionID int32, offset int64, err error) {
+	serviceLogURL := httputils.SetHTTPPrefix(producer.Configuration.URL)
+
+	client := &http.Client{
+		Timeout: time.Second * producer.Configuration.Timeout,
+	}
+
+	req, err := http.NewRequest(http.MethodPost, serviceLogURL, bytes.NewBuffer(msg))
+	req.Header.Add("Authorization", "Bearer "+producer.AccessToken)
+	if err != nil {
+		log.Error().Msgf("Got error while setting up HTTP request -  %s", err.Error())
+		return -1, -1, err
+	}
+
+	response, err := client.Do(req)
+	if err != nil {
+		log.Error().Msgf("Got error while making the HTTP request - %s", err.Error())
+		return -1, -1, err
+	}
+
+	if response.StatusCode == http.StatusUnauthorized {
+		err = producer.refreshToken()
+		if err != nil {
+			log.Error().Msg("Access token could not be refreshed")
+			return -1, -1, err
+		}
+		return producer.ProduceMessage(msg)
+	} else if response.StatusCode != http.StatusCreated {
+		log.Error().Msgf("Service log could not process the message - received a response with status '%s'", response.Status)
+		return -1, -1, errors.New("service log could not process the message")
+	}
+
+	return 0, 0, nil
+}
+
+// Close closes Producer (in case of Service Log implementation, it does not do anything)
+func (producer *Producer) Close() error {
+	return nil
+}
+
+func (producer *Producer) refreshToken() error {
+	params := url.Values{}
+	params.Add("grant_type", `refresh_token`)
+	params.Add("client_id", `rhsm-api`)
+	params.Add("refresh_token", producer.Configuration.OfflineToken)
+	body := strings.NewReader(params.Encode())
+
+	tokenRefreshmentURL := httputils.SetHTTPPrefix(producer.Configuration.TokenRefreshmentURL)
+	req, err := http.NewRequest(http.MethodPost, tokenRefreshmentURL, body)
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	if err != nil {
+		log.Error().Msgf("Got error while setting up HTTP request for refreshing access token -  %s", err.Error())
```

Comment:
It's better not to include the error in the `.Msg()`. It's better to use `log.Error().Err(err).Msg(...)`.

Reply:
Should be resolved now